src/test/testLogger.ts

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { isCI } from '../client/common/constants';
+import { LogLevel } from '../client/logging/levels';
+import { configureLogger, createLogger, getPreDefinedConfiguration, logToAll } from '../client/logging/logger';
+
+const monkeyPatchLogger = createLogger();
+
+export function initializeLogger() {
+    const config = getPreDefinedConfiguration();
+    if (isCI && process.env.VSC_PYTHON_LOG_FILE) {
+        delete config.console;
+        // This is a separate logger that matches our config but
+        // does not do any console logging.
+        configureLogger(monkeyPatchLogger, config);
+        // Send console.*() to the non-console loggers.
+        monkeypatchConsole();
+    }
+}
+
+// Ensure that the console functions are bound before monkeypatching.
+import '../client/logging/transports';
+
+/**
+ * What we're doing here is monkey patching the console.log so we can
+ * send everything sent to console window into our logs.  This is only
+ * required when we're directly writing to `console.log` or not using
+ * our `winston logger`.  This is something we'd generally turn on, only
+ * on CI so we can see everything logged to the console window
+ * (via the logs).
+ */
+function monkeypatchConsole() {
+    // The logging "streams" (methods) of the node console.
+    const streams = ['log', 'error', 'warn', 'info', 'debug', 'trace'];
+    const levels: { [key: string]: LogLevel } = {
+        error: LogLevel.Error,
+        warn: LogLevel.Warn
+    };
+    // tslint:disable-next-line:no-any
+    const consoleAny: any = console;
+    for (const stream of streams) {
+        // Using symbols guarantee the properties will be unique & prevents
+        // clashing with names other code/library may create or have created.
+        // We could use a closure but it's a bit trickier.
+        const sym = Symbol.for(stream);
+        consoleAny[sym] = consoleAny[stream];
+        // tslint:disable-next-line: no-function-expression
+        consoleAny[stream] = function () {
+            const args = Array.prototype.slice.call(arguments);
+            const fn = consoleAny[sym];
+            fn(...args);
+            const level = levels[stream] || LogLevel.Info;
+            logToAll([monkeyPatchLogger], level, args);
+        };

```

Comment:
Closures can be tricky.  `stream` here **might** be bound to `trace` in all functions since that is the last value from the last iteration of the "for" loop.  At least that's how closures are bound in Python.  I'm pretty sure that's how they are bound in Javascript too.  (The same applies to `sym`.)  So you will want to double-check.  (I realize this is basically code *I* wrote, but I may have gotten it wrong.)

If you find the closures are binding the wrong values then you will need to force binding them correctly (e.g. at the end of the loop.  I'm pretty sure using `Function.bind()` will do the right thing:

```suggestion
        consoleAny[stream] = function (symClosure, streamClosure) {
            const args = Array.prototype.slice.call(arguments, 2);
            const fn = consoleAny[symClosure];
            fn(...args);
            const level = levels[streamClosure] || LogLevel.Info;
            logToAll([monkeyPatchLogger], level, args);
        };
        consoleAny[stream] = consoleAny[stream].bind(this, sym, stream);
```

Reply:
Can you help with how to check if `stream` here is bound to `trace`?

Reply:
Here's a minimal example:

```typescript
const values = [1, 2. 3];

const badClosures = [];
for (const v of values) {
    badClosures.push(function () {
        console.log(v);
    });
}
for (const fn of badClosures) {
    fn();
}
// 3
// 3
// 3

const goodClosures = [];
for (const v of values) {
    goodClosures.push(function () {
        console.log(v);
    }.bind(this, v));
}
for (const fn of goodClosures) {
    fn();
}
// 1
// 2
// 3
```

Reply:
I've checked that both cases print 1 2 3, so this shouldn't be a problem.

Reply:
Thanks for the example btw.